include/swift/Sema/Constraint.h

```diff
@@ -223,7 +223,9 @@ enum class ConstraintKind : char {
   ///
   /// Binds the RHS type to a tuple of the params of a function typed LHS. Note
   /// this discards function parameter flags.
-  BindTupleOfFunctionParams
+  BindTupleOfFunctionParams,
+  /// The first type is a type pack, and the second type is its reduced shape.
+  ShapeOf,
```

Comment:
Wouldn't `EqualShape` constraint be easier to work with instead of having to go through an additional type variable? Tailoring diagnostics might be more challenging with this approach as well because the mismatch would be against a regular `Bind` constraint.

Reply:
Yep, I realized the same last night. I'll do it in the next PR